File: src/classes/JointPlanner.m
classdef JointPlanner < handle
    
    properties
        jpos
        tf
        nj
        ntraj
        segplanner
        plan_idx
        unplan_idx
        seg_idx
        t
        cycle_time
        plan_completed

        sync
    end

    methods
        function obj = JointPlanner(q0,sync,cycle_time)
            obj.jpos = q0;
            obj.sync = sync;
            obj.nj = length(q0);
            obj.plan_idx = [];
            obj.unplan_idx = [];
            obj.seg_idx = 1;
            obj.t = 0;
            obj.cycle_time = cycle_time;
            obj.plan_completed = false;
        end

        function AddJntPos(obj,q,jvmax,jamax)
            obj.jpos = [obj.jpos, q];
            obj.ntraj = size(obj.jpos,2)-1;
            for traj_idx=1:obj.ntraj
                q0 = obj.jpos(:,traj_idx);
                qf = obj.jpos(:,traj_idx+1);
                planidx = []; unplanidx = [];
                for idx=1:obj.nj
                    if(abs(q0(idx)-qf(idx))>1e-5)
                        planidx = [planidx,idx];
                    else
                        unplanidx = [unplanidx,idx];
                    end
                end
                tf_tmp = [];
                for idx=planidx
                    uplanner{idx} = LspbPlanner([q0(idx),qf(idx)],...
                                            jvmax(idx),jamax(idx));
                    tf_tmp = [tf_tmp,uplanner{idx}.tf];
                end
                obj.tf(traj_idx) = max(tf_tmp);
                if(obj.sync)
                    for idx=planidx
                        uplanner{idx} = LspbPlanner([q0(idx),qf(idx)],...
                                                jvmax(idx),jamax(idx),obj.tf(traj_idx));
                    end
                end
                obj.segplanner{traj_idx} = uplanner;
                obj.plan_idx{traj_idx} = planidx;
                obj.unplan_idx{traj_idx} = unplanidx;
            end
        end
        
        function Stop(obj,jp,jv,jvmax,jamax)
            obj.jpos = jp;
            obj.ntraj = 1;
            obj.segplanner = {};
            obj.plan_idx = {};
            obj.unplan_idx = {};
            planidx = []; unplanidx = [];
            for idx=1:obj.nj
                if abs(jv(idx))>1e-5
                    planidx = [planidx,idx];
                else
                    unplanidx = [unplanidx,idx];
                end
            end
            tf_tmp = [];
            for idx=planidx
                uplanner{idx} = LspbPlanner(jp(idx),jvmax(idx),jamax(idx),[],[jv(idx),0]);
                tf_tmp = [tf_tmp,uplanner{idx}.ta];
            end
            obj.tf = max(tf_tmp);
            if obj.sync
                for idx=planidx
                    uplanner{idx} = LspbPlanner(jp(idx),jvmax(idx),jamax(idx),obj.tf,[jv(idx),0]);
                end
            end
            obj.segplanner{1} = uplanner;
            obj.plan_idx{1} = planidx;
            obj.unplan_idx{1} = unplanidx;
            
            obj.plan_completed = false;
            obj.t = 0;
            obj.seg_idx = 1;
        end

        function [p,v,a] = GenerateMotion(obj)
            [p,v,a] = obj.GenerateSegMotion(obj.seg_idx,obj.t);
            obj.t = obj.t+obj.cycle_time;       
            if (obj.seg_idx==obj.ntraj) && (abs(obj.t-obj.tf(obj.seg_idx))<obj.cycle_time)
                obj.plan_completed = true;
            else
                if (abs(obj.t-obj.tf(obj.seg_idx))<obj.cycle_time)
                    obj.seg_idx  = obj.seg_idx + 1;
                    obj.t = 0;
                end
            end
        end
        
        function [pos,vel,acc] = GenerateTraj(obj,dt)
            pos = []; vel = []; acc = [];
            for traj_idx=1:obj.ntraj
                for t=0:dt:obj.tf(traj_idx)
                    [p,v,a] = obj.GenerateSegMotion(traj_idx,t);
                    pos = [pos,p]; vel = [vel,v]; acc = [acc, a];
                end
            end
        end
        
        function [p,v,a] = GenerateSegMotion(obj,traj_idx,t)
            p = zeros(obj.nj,1); v = zeros(obj.nj,1); a = zeros(obj.nj,1);
            if ~isempty(obj.plan_idx{traj_idx})
                for idx=obj.plan_idx{traj_idx}
                    t_limit(idx) = LimitNumber(0,t,obj.segplanner{traj_idx}{idx}.tf);
                    [p(idx),v(idx),a(idx)] = obj.segplanner{traj_idx}{idx}.GenerateMotion(t_limit(idx));
                end
            end
            if ~isempty(obj.unplan_idx{traj_idx})
                for idx=obj.unplan_idx{traj_idx}
                    p(idx) = obj.jpos(idx,traj_idx);
                end
            end
        end



    end




end

File: src/cfiles/JointPlanner.cpp
#include "JointPlanner.h"

JointPlanner::JointPlanner(Vector6d q0, bool sync)
	:_sync(sync)
{
	_jpos.push_back(q0);
	_t = 0;
	_plan_completed = false;
	_vmax = g_jvmax;
	_amax = g_jamax;
	_ntraj = 0;
	_traj_idx = 0;
}

void JointPlanner::SetKineConstraint(double* vmax, double* amax)
{
	_vmax = vmax;
	_amax = amax;
}

void JointPlanner::AddViaPos(MatrixXd* jpos)
{
	_ntraj = (int)jpos->cols();
	for (int traj_idx = 0; traj_idx<_ntraj; traj_idx++)
	{
		_jpos.push_back(jpos->col(traj_idx));
		Vector6d q0, qf;
		q0 = _jpos[traj_idx];
		qf = _jpos[traj_idx+1];
		vector<int> planidx, unplanidx;
		for (int jidx = 0; jidx<6; jidx++)
		{
			if (fabs(q0(jidx)-qf(jidx))>EPS3)
				planidx.push_back(jidx);
			else
				unplanidx.push_back(jidx);
		}
		_jplanned.push_back(planidx);
		_junplanned.push_back(unplanidx);
		vector<double> tf_jnt;
		vector<LspbPlanner*> vec_uplanner;
		vec_uplanner.resize(6);
		for (int jidx : _jplanned[traj_idx])
		{
			if ((qf(jidx)-g_qlimit_ur[2*jidx+1]>0)||(qf(jidx)-g_qlimit_ur[2*jidx]<0))
				throw eErrJointOutOfRange;

			LspbPlanner* uplanner = new LspbPlanner(Vector2d(q0(jidx), qf(jidx)),
				_vmax[jidx], _amax[jidx]);
			vec_uplanner[jidx] = uplanner;
			tf_jnt.push_back(uplanner->GetFinalTime());
		}
		if (tf_jnt.size()>0)
			_tf.push_back(*max_element(tf_jnt.begin(), tf_jnt.end()));
		else
			_tf.push_back(0);
		if (_sync)
		{
			for (int jidx : _jplanned[traj_idx])
			{
				LspbPlanner* uplanner = new LspbPlanner(Vector2d(q0(jidx), qf(jidx)),
					_vmax[jidx], _amax[jidx], _tf[traj_idx]);
				vec_uplanner[jidx] = uplanner;
			}
		}
		_segplanner.push_back(vec_uplanner);
	}
}						   

void JointPlanner::GeneratePath(Vector6d& jpos)
{
	RobotTools::JAVP javp;
	//generate joint position
	if (_jplanned[_traj_idx].size()>0)
	{
		for (int jidx:_jplanned[_traj_idx])
		{
			javp = _segplanner[_traj_idx][jidx]->GenerateMotion(_t);
			jpos(jidx) = javp.pos;
		}
		_t += g_cycle_time;
		MathTools::LimitMax(_tf[_traj_idx], _t);
	}
	if (_junplanned[_traj_idx].size()>0)
	{
		for (int jidx : _junplanned[_traj_idx])
			jpos(jidx) = _jpos[_traj_idx](jidx);
	}
	//logical judgment about trajectory index and plan status
	if ((_traj_idx==_ntraj-1)&&(fabs(_t-_tf[_traj_idx])<EPS3))
		_plan_completed = true;
	else
	{
		if (fabs(_t-_tf[_traj_idx])<EPS3)
		{
			_traj_idx += 1;
			_t = 0;
		}
	}
}

void JointPlanner::GenerateMotion(Vector6d& jpos, Vector6d& jvel, Vector6d& jacc)
{
	RobotTools::JAVP javp;
	//generate joint position
	if (_jplanned[_traj_idx].size()>0)
	{
		for (int jidx : _jplanned[_traj_idx])
		{
			javp = _segplanner[_traj_idx][jidx]->GenerateMotion(_t);
			jpos(jidx) = javp.pos;
			jvel(jidx) = javp.vel;
			jacc(jidx) = javp.acc;
			_t += g_cycle_time;
			MathTools::LimitMax(_tf[_traj_idx], _t);
		}
	}
	if (_junplanned[_traj_idx].size()>0)
	{
		for (int jidx : _junplanned[_traj_idx])
		{
			jpos(jidx) = _jpos[_traj_idx](jidx);
			jvel(jidx) = 0;
			jacc(jidx) = 0;
		}
	}
	//logical judgment about trajectory index and plan status
	if ((_traj_idx==_ntraj-1)&&(fabs(_t-_tf[_traj_idx])<EPS3))
		_plan_completed = true;
	else
	{
		if (fabs(_t-_tf[_traj_idx])<EPS3)
		{
			_traj_idx += 1;
			_t = 0;
		}
	}

}

void JointPlanner::Reset(Vector6d q0, bool sync)
{
	ClearTmp();
	_jpos.clear();
	_jpos.push_back(q0);
	_tf.clear();
	_jplanned.clear();
	_junplanned.clear();

	_plan_completed = false;
	_t = 0;
	_sync = sync;
	_ntraj = 0;
	_traj_idx = 0;
}

void JointPlanner::ClearTmp()
{
	for (int traj_idx = 0; traj_idx<_ntraj; traj_idx++)
	{
		for (auto it = _segplanner[traj_idx].begin(); it!=_segplanner[traj_idx].end(); it++)
		{
			if (*it!=nullptr)
			{
				delete *it;
				*it = nullptr;
			}
		}
		_segplanner[traj_idx].clear();
	}
	_segplanner.clear();
}

